lib/auth/client_tls_config_generator.go

```diff
@@ -87,6 +87,14 @@ func NewClientTLSConfigGenerator(cfg ClientTLSConfigGeneratorConfig) (*ClientTLS
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
+	// we're going to start the generator, which means that we are potentially
+	// going to read from cfg.TLS at any point after returning, which means that
+	// either the caller must always give us a cloned [tls.Config], or we just
+	// make one here
+	cfg.TLS = cfg.TLS.Clone()
```

Comment:
i don't know much about this tls config generator, but i'm curious, is there any potential correctness issue here? We are trying to set GetConfigForClient elsewhere which caused this race, is it wrong for this tls.Config to stick around without that GetConfigForClient?

Reply:
It is (and with this change we're effectively restoring the behavior that we had before #43937). The `GetConfigForClient` callback doesn't have an intrinsic way to access the "intended" `tls.Config` (much to @fspmarshall and I's chagrin) so any further changes to the original `tls.Config` would be lost anyway, and the `GetConfigForClient` in the `tls.Config` returned by `GetConfigForClient` is ignored, so this is correct as long as only `GetConfigForClient` is modified (which is all you can modify to begin with).

Reply:
(and that's the only field we modify after each call to `NewClientTLSConfigGenerator` currently in master)